ZEngine/src/GlfwWindow/VulkanWindow.cpp

```diff
@@ -420,10 +439,89 @@ namespace ZEngine::Window::GLFWWindow
 
     void VulkanWindow::Render()
     {
+        auto&       current_frame     = this->GetCurrentWindowFrame();
+        const auto& performant_device = Engine::GetVulkanInstance()->GetHighPerformantDevice();
+        auto        device_handle     = performant_device.GetNativeDeviceHandle();
+        auto        present_queue     = performant_device.GetCurrentGraphicQueueWithPresentSupport().Queue;
+
+        VkResult wait_for_fences_result = vkWaitForFences(device_handle, 1, &(current_frame.Fence), VK_TRUE, UINT64_MAX); // wait indefinitely instead of periodically checking
+        if (wait_for_fences_result == VK_TIMEOUT)
+        {
+            return;
+        }
+
+        ZENGINE_VALIDATE_ASSERT(vkResetFences(device_handle, 1, &current_frame.Fence) == VK_SUCCESS, "Failed to reset Fence semaphore")
+
+        uint32_t image_index;
+        VkResult acquire_image_result = vkAcquireNextImageKHR(device_handle, m_swapchain, UINT64_MAX, current_frame.ImageAcquiredSemaphore, VK_NULL_HANDLE, &image_index);
+        if (acquire_image_result == VK_ERROR_OUT_OF_DATE_KHR || acquire_image_result == VK_SUBOPTIMAL_KHR || acquire_image_result == VK_NOT_READY || acquire_image_result == VK_TIMEOUT)
+        {
+            return;
+        }
+        ZENGINE_VALIDATE_ASSERT(acquire_image_result == VK_SUCCESS, "Failed to acquire image from Swapchain")
+
+        if (!current_frame.GraphicCommandBuffers.empty())
+        {
+            vkFreeCommandBuffers(
```

Comment:
Technically, you could ommit this, since you also call vkResetCommandPool on the whole pool, which resets all allocated command buffers from that pool back to the initial state. So you can immediately call vkBeginCommandBuffer on any of them again to record new commands without freeing and reallocating them first.

For completeness: Another option would be to not use vkResetCommandPool, but reset the individual command buffers using vkResetCommandBuffer and record into them again with vkBeginCommandBuffer. You still won't need to free them this way. Caveat: this requires the VK_COMMAND_POOL_CREATE_RESET_COMMAND_BUFFER_BIT flag during the CommandPool's creation.

Reply:
Thanks a lot !,
I think this is no longer applicable, given the fact I revisited the whole Vulkan resource management.
we might need to review it again  -- sorry for this huge PR 😅